Src/FluentAssertions/Specialized/TaskCompletionSourceAssertions.cs

```diff
@@ -8,6 +8,105 @@ namespace FluentAssertions.Specialized;
 
 #pragma warning disable CS0659 // Ignore not overriding Object.GetHashCode()
 #pragma warning disable CA1065 // Ignore throwing NotSupportedException from Equals
+
+#if NET6_0_OR_GREATER
+
+public class TaskCompletionSourceAssertions
+{
+    private readonly TaskCompletionSource subject;
+
+    public TaskCompletionSourceAssertions(TaskCompletionSource tcs)
+        : this(tcs, new Clock())
+    {
+    }
+
+    public TaskCompletionSourceAssertions(TaskCompletionSource tcs, IClock clock)
+    {
+        subject = tcs;
+        Clock = clock ?? throw new ArgumentNullException(nameof(clock));
+    }
+
+    private protected IClock Clock { get; }
+
+    /// <summary>
+    /// Asserts that the <see cref="Task"/> of the current <see cref="TaskCompletionSource{T}"/> will complete within the specified time.
+    /// </summary>
+    /// <param name="timeSpan">The allowed time span for the operation.</param>
+    /// <param name="because">
+    /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+    /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+    /// </param>
+    /// <param name="becauseArgs">
+    /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+    /// </param>
+    public async Task CompleteWithinAsync(
```

Comment:
When the TCS is completed, there is no addition information available to be asserted.
I think it's too theoretical to be useful.
If a user is that creative it can be easily changed without breaking change.

Reply:
> If a user is that creative it can be easily changed without breaking change.

Because the return type is not part of the method signature?

Reply:
Because the method currently does not provide a (real) return value and it can be added easily.

Reply:
Ok, because of consistency with the `Task` assertions I'll return `AndContstraint`.

Reply:
Now, I do not know anymore what is the right and consistent way. Working on `NotCompletedWithinAsync` for `Task` in #1967 it was looking absolutely consistent to return `AndConstraint`. Now, I'm unsure because we have options now:
1. Returning `AndConstraint` in the new assertions for non-generic TCS, leaving the existing one to not support chaining. Then we have inconsistency in the two variants for TCS.
2. Returning `AndConstraint` in the new assertions and changing the existing implementation to also return `AndConstraint`.
3. Returning just `Task` in the new assertion similar to the existing one, the generic one.

Reply:
If 1. is correctly read as the following, I would go with that.

```cs
public class TaskCompletionSourceAssertions
{
    public Task<AndConstraint<TaskCompletionSourceAssertions>> CompleteWithinAsync(TimeSpan timeSpan, string because = "", params object[] becauseArgs) { }
    public Task<AndConstraint<TaskCompletionSourceAssertions>> NotCompleteWithinAsync(TimeSpan timeSpan, string because = "", params object[] becauseArgs) { }
}
```

Reply:
Hey @jnyrup, I started rewriting as you mentioned.
I'm unable to fix the test `Should_methods_have_a_matching_overload_to_guard_against_chaining_and_constraints`. The error message is confusing because the (new) base class `TaskCompletionSourceAssertionsBase` is mentioned while the assertion class itself has received a fake `Should`. Please help to understand your test.

Further, I'm unsure whether `TaskCompletionSourceAssertions` needs to be added manually to `realOverloads`, and why `DateTimeRangeAssertions` and `DateTimeOffsetRangeAssertions` are added manually.

Please check my SAVEPOINT.

Reply:
See https://github.com/jnyrup/fluentassertions/commit/6dd6676e4195b9bdae7c1a7f0856de79144c96de for code fixes and comments

Reply:
Ok, I guess I understand the problems.
I added two comments there.

Reply:
I can't see the comments on mobile, so commenting here. 

The Base class does not strictly have to be abstract, I usually just go with abstract if it does not make sense to create an instance of a class. 

I think they should be private protected.